libs/ui/src/unlock_machine_screen.tsx

```diff
@@ -39,43 +43,67 @@ const EnteredCode = styled.div`
 
 type CheckingPassCodeAuth =
   | DippedSmartcardAuth.CheckingPasscode
+  | DippedSmartCardAuth.CheckingPin
   | InsertedSmartcardAuth.CheckingPasscode;
 
 interface Props {
   auth: CheckingPassCodeAuth;
   grayBackground?: boolean;
+  checkPin?: (pin: string) => Promise<void>;
 }
 
 export function UnlockMachineScreen({
   auth,
   grayBackground,
+  checkPin: checkPinFromProps,
 }: Props): JSX.Element {
   assert(auth.status === 'checking_passcode');
 
   const [currentPasscode, setCurrentPasscode] = useState('');
-  const handleNumberEntry = useCallback((digit: number) => {
-    setCurrentPasscode((prev) =>
-      `${prev}${digit}`.slice(0, SECURITY_PIN_LENGTH)
-    );
-  }, []);
+
+  const handleNumberEntry = useCallback(
+    async (digit: number) => {
+      const passcode = `${currentPasscode}${digit}`.slice(
+        0,
+        SECURITY_PIN_LENGTH
+      );
+      setCurrentPasscode(passcode);
+      if (passcode.length === SECURITY_PIN_LENGTH) {
+        if ('checkPasscode' in auth) {
+          auth.checkPasscode(passcode);
+        } else {
+          assert(checkPinFromProps !== undefined);
+          await checkPinFromProps(passcode);
+        }
+        setCurrentPasscode('');
+      }
+    },
+    [auth, checkPinFromProps, currentPasscode]
+  );
+
   const handleBackspace = useCallback(() => {
     setCurrentPasscode((prev) => prev.slice(0, -1));
   }, []);
+
   const handleClear = useCallback(() => {
     setCurrentPasscode('');
   }, []);
 
   useEffect(() => {
-    if (isFeatureFlagEnabled(BooleanEnvironmentVariableName.SKIP_PIN_ENTRY)) {
-      auth.checkPasscode(auth.user.passcode);
-      return;
+    async function bypassPinEntry() {
+      if ('checkPasscode' in auth) {
+        auth.checkPasscode(auth.user.passcode);
+      } else {
+        assert(checkPinFromProps !== undefined);
+        await checkPinFromProps(auth.user.passcode);
+      }
     }
-
-    if (currentPasscode.length === SECURITY_PIN_LENGTH) {
-      auth.checkPasscode(currentPasscode);
-      setCurrentPasscode('');
+    if (isFeatureFlagEnabled(BooleanEnvironmentVariableName.SKIP_PIN_ENTRY)) {
+      void bypassPinEntry();
     }
-  }, [currentPasscode, auth]);
+    // Run this hook once on mount
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
```

Comment:
Yes that's the plan! In fact, I plan on removing PINs from the user objects exposed to the frontend, so skipping PIN entry will have to move to the backend

I just can't make these changes yet since `UnlockMachineScreen` is a shared component still in use by machines with frontend auth